syft/pkg/cataloger/java/graalvm_native_image_cataloger.go

```diff
@@ -0,0 +1,316 @@
+	data, err := dataSection.Data()
+	if err != nil {
+		log.Warnf("native-image cataloger: cannot read the .data section.")
+		return nil, err
+	}
+	sbomLocation := sbom.Value - dataSectionBase
+	lengthLocation := sbomLength.Value - dataSectionBase
+
+	return decompressSbom(data, sbomLocation, lengthLocation)
+}
+
+// FetchPkgs obtains the packages from a Native Image given as a Mach O file.
+func (ni NativeImageMachO) FetchPkgs() ([]pkg.Package, error) {
+	var sbom macho.Symbol
+	var sbomLength macho.Symbol
+	var svmVersion macho.Symbol
+
+	bi := ni.File
+	for _, s := range bi.Symtab.Syms {
+		switch s.Name {
+		case "_" + nativeImageSbomSymbol:
+			sbom = s
+		case "_" + nativeImageSbomLengthSymbol:
+			sbomLength = s
+		case "_" + nativeImageSbomVersionSymbol:
+			svmVersion = s
+		}
+	}
+	if sbom.Value == 0 || sbomLength.Value == 0 || svmVersion.Value == 0 {
+		log.Infof("native-image cataloger: %v.", nativeImageMissingSymbolsError)
+		return nil, errors.New(nativeImageMissingSymbolsError)
+	}
+
+	dataSegment := bi.Segment("__DATA")
+	if dataSegment == nil {
+		return nil, nil
+	}
+	databuf, err := dataSegment.Data()
+	if err != nil {
+		log.Warnf("native-image cataloger: cannot obtain buffer from data segment.")
+		return nil, nil
+	}
+	sbomLocation := sbom.Value - dataSegment.Addr
+	lengthLocation := sbomLength.Value - dataSegment.Addr
+
+	return decompressSbom(databuf, sbomLocation, lengthLocation)
+}
+
+// fetchPkgs provides the packages available in a UnionReader.
+func fetchPkgs(reader unionreader.UnionReader, filename string) []pkg.Package {
+	var pkgs []pkg.Package
+	imageformats := []func(string, io.ReaderAt) (NativeImage, error){newElf, newMachO}
+
+	// NOTE: multiple readers are returned to cover universal binaries, which are files
+	// with more than one binary
+	readers, err := unionreader.GetReaders(reader)
+	if err != nil {
+		log.Warnf("native-image cataloger: failed to open a binary: %v", err)
+		return nil
+	}
+	for _, r := range readers {
+		for _, make := range imageformats {
+			ni, err := make(filename, r)
+			if err != nil {
+				continue
+			}
+			newpkgs, err := ni.FetchPkgs()
+			if err != nil {
+				log.Warnf("native-image cataloger: error extracting SBOM from %s %v", filename, err)
+				continue
+			}
+			pkgs = append(pkgs, newpkgs...)
+		}
+	}
+	return pkgs
+}
+
+// Catalog attempts to find any native image executables reachable from a resolver.
+func (c *NativeImageCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package, []artifact.Relationship, error) {
+	var pkgs []pkg.Package
+	fileMatches, err := resolver.FilesByMIMEType(internal.ExecutableMIMETypeSet.List()...)
+	if err != nil {
+		return pkgs, nil, fmt.Errorf("failed to find binaries by mime types: %w", err)
+	}
+
+	for _, location := range fileMatches {
+		readerCloser, err := resolver.FileContentsByLocation(location)
+		if err != nil {
+			log.Infof("native-image cataloger: opening file: %v", err)
```

Comment:
nit: should this be "error opening file"? and either `log.Warnf` (would be displayed) or `log.Debugf` (would only be displayed with `-vv` flag)?

Reply:
agreed with this, there are a lot of Info log lines for handling error cases. Warn is great in cases where there is something wrong that warrants the user to take a look and report an issue in all of cases. Debug is better when the error may occur occasionally or doesn't warrant the user to take action.

Reply:
This makes more sense, we can rely on Debug when we want to log errors that do not need to reach users, but may be helpful for development/understanding the cataloger's logic.